server/src/main/java/org/elasticsearch/index/mapper/blockloader/docvalues/DenseVectorBlockLoader.java

```diff
@@ -46,43 +59,60 @@ public AllReader reader(LeafReaderContext context) throws IOException {
                     if (fieldType.isNormalized()) {
                         NumericDocValues magnitudeDocValues = context.reader()
                             .getNumericDocValues(fieldType.name() + COSINE_MAGNITUDE_FIELD_SUFFIX);
-                        return new FloatDenseVectorNormalizedValuesBlockReader(floatVectorValues, dimensions, magnitudeDocValues);
+                        return new FloatDenseVectorNormalizedValuesBlockReader<>(
+                            floatVectorValues,
+                            dimensions,
+                            processor,
+                            magnitudeDocValues
+                        );
                     }
-                    return new FloatDenseVectorValuesBlockReader(floatVectorValues, dimensions);
+                    return new FloatDenseVectorValuesBlockReader<>(floatVectorValues, dimensions, processor);
                 }
             }
             case BYTE -> {
                 ByteVectorValues byteVectorValues = context.reader().getByteVectorValues(fieldName);
                 if (byteVectorValues != null) {
-                    return new ByteDenseVectorValuesBlockReader(byteVectorValues, dimensions);
+                    return new ByteDenseVectorValuesBlockReader<>(byteVectorValues, dimensions, processor);
                 }
             }
             case BIT -> {
                 ByteVectorValues byteVectorValues = context.reader().getByteVectorValues(fieldName);
                 if (byteVectorValues != null) {
-                    return new BitDenseVectorValuesBlockReader(byteVectorValues, dimensions);
+                    return new BitDenseVectorValuesBlockReader<>(byteVectorValues, dimensions, processor);
                 }
             }
         }
 
         return new ConstantNullsReader();
     }
 
-    private abstract static class DenseVectorValuesBlockReader<T extends KnnVectorValues> extends BlockDocValuesReader {
+    /**
+     * Abstract base class for readers that process dense vector values.
+     * Provides common iteration and null-handling logic.
+     *
+     * @param <T> The type of KNN vector values (FloatVectorValues, ByteVectorValues, etc.)
+     * @param <B> The type of builder used to construct blocks
+     */
+    private abstract static class AbstractVectorValuesReader<T extends KnnVectorValues, B extends BlockLoader.Builder> extends
+        BlockDocValuesReader {
+
         protected final T vectorValues;
         protected final KnnVectorValues.DocIndexIterator iterator;
+        protected final DenseVectorBlockLoaderProcessor<B> processor;
```

Comment:
FOR LATER: I think it'd probably be faster to have totally separate subclasses of `ColumnAtATimeReader` for each similarity rather than make the megamorphic calls to them. But let's get this in as you proposed it and then play with it.